### DISCUSSION TITLE
Welcome to FreeCAD_Utils Discussions!

### DISCUSSION THREAD

**Discussion by mnesarco** [mnesarco/FreeCAD_Utils#11] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by macdroid53** [comment]
Hi, running utilities in 0.22dev and get this eror:
`PropertyPythonObject::toString(): failed for <class 'freecad.mnesarco.scripts.script.ScriptObjectGui'>
09:51:08  pyException: Traceback (most recent call last):
  File "/usr/lib/python3.11/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
<class 'TypeError'>: Object of type ViewProviderDocumentObject is not JSON serializable`

Also, probably showing my limited Python skills...

In the example Python it says:
`# global variable TIME contains the current Tick.

print(TIME)
`

But, TIME doesn't seem to be a default global.

**Reply by mnesarco** [reply]
Hi @macdroid53 , could you provide a reproducible case step by step?

**Reply by macdroid53** [reply]
The tree view shows the contents of the document, the heyworld branch contains what the process of creating it puts there (the demo code) . Hit save, get this:
`10:49:40  PropertyPythonObject::toString(): failed for <class 'freecad.mnesarco.scripts.script.ScriptObjectGui'>
10:49:40  pyException: Traceback (most recent call last):
  File "/usr/lib/python3.11/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
<class 'TypeError'>: Object of type ViewProviderDocumentObject is not JSON serializable
`
As for TIME, use Diameter or Angle in an expression, get 0.0.

**Reply by mnesarco** [reply]
I didn't understand a word :D of what you said, but i will install 0.22dev and try to reproduce.

**Reply by macdroid53** [reply]
What part do I need to make clearer?

**Reply by mnesarco** [reply]
Let me check the weekly build by my self, if the error is there it will show by itself.

**Reply by mnesarco** [reply]
I suppose that you are using the ScriptObject from mnesarco_utils, so tried to reproduce doing this:

1. Create a new document
2. Save the document
3. Add a Script Object from mnesarco_utils
4. Set the name
5. Left the example code in it
6. Click ok

until now, no errors so far

7. Then i set scale to 5 (for testing)
8. Nothing happens
9. Then i save the document
10. Nothing happens
11. Close FreeCAD
12. Open FreeCAD
13. Reopen the saved file
14. Nothing happens

Maybe I am doing something different?

**Reply by macdroid53** [reply]
Create a new document
    Save the document
    Add a Script Object from mnesarco_utils
    Set the name
    Left the example code in it
    Click ok

No problem.

Create a Body/Sketch w/ rectangle
Create a horizontal constraint, choose expression, enter "<name of Python object>.Diameter"
Constraint value is 0.0
Attempt to save file:

`11:29:41  Updating geometry: Error build geometry(1): Both points are equal
11:29:41  Invalid solution from DogLeg solver.`

Attempt to save file:

`11:30:28  PropertyPythonObject::toString(): failed for <class 'freecad.mnesarco.scripts.script.ScriptObjectGui'>
11:30:28  pyException: Traceback (most recent call last):
  File "/usr/lib/python3.11/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
<class 'TypeError'>: Object of type ViewProviderDocumentObject is not JSON serializable`

**Reply by macdroid53** [reply]
And there doesn't seem to be a global variable TIME as mentioned in the code example.

**Reply by mnesarco** [reply]
The first error is because your Diameter is zero (0). In the example code, Diameter is a function of time, so set time to something greater than zero in the property editor of the Script Object and press Ctrl+R to refresh.

I saved the document without any error, so we should have something different. I will explore more tomorrow.

**Reply by mnesarco** [reply]
**TIME** is injected by mnesarco_utils, it is there and you can set it from the properties editor manually or using a Timer object and binding it.

**Reply by macdroid53** [reply]
I had not noticed the TIME property. Yes, setting it fixes the sketch error.

But, saving the file still results in:
`11:51:34  PropertyPythonObject::toString(): failed for <class 'freecad.mnesarco.scripts.script.ScriptObjectGui'>
11:51:34  pyException: Traceback (most recent call last):
  File "/usr/lib/python3.11/json/__init__.py", line 231, in dumps
    return _default_encoder.encode(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 200, in encode
    chunks = self.iterencode(o, _one_shot=True)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 258, in iterencode
    return _iterencode(o, 0)
           ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/json/encoder.py", line 180, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
<class 'TypeError'>: Object of type ViewProviderDocumentObject is not JSON serializable
`

**Reply by mnesarco** [reply]
I could not reproduce it. I tried to upload a video but Github did not play it properly.

**Reply by mnesarco** [reply]
Video: https://www.youtube.com/watch?v=cQHcqNUzcxM

**Reply by macdroid53** [reply]
Thought it might be because I was using a PD Body/Sketch.

Still can't save.

Are you using Python 3.11?

**Reply by macdroid53** [reply]
Wait, can't save is wrong the save works, FC just throws the errors shown.

**Reply by mnesarco** [reply]
Apparently the issue is not specific to Mnesarco_Utils:

https://forum.freecad.org/viewtopic.php?p=736727&hilit=is+not+JSON+serializable#p736727
